lib/Dialect/Seq/Transforms/LowerSeqToSV.cpp

```diff
@@ -73,10 +80,378 @@ struct CompRegLower : public OpConversionPattern<CompRegOp> {
+        })
+        .Case<hw::StructType>([&](hw::StructType s) {
+          for (auto elem : s.getElements())
+            recurse(elem.type, member + "." + elem.name.getValue());
+        })
+        .Default([&](auto type) { emitRandomInit(reg, type, member); });
+  };
+  recurse(type, "");
+}
+
+void FirRegLower::addToAlwaysBlock(sv::EventControl clockEdge, Value clock,
+                                   std::function<void(OpBuilder &)> body,
+                                   ::ResetType resetStyle,
+                                   sv::EventControl resetEdge, Value reset,
+                                   std::function<void(OpBuilder &)> resetBody) {
+  auto loc = module.getLoc();
+  auto builder =
+      ImplicitLocOpBuilder::atBlockTerminator(loc, module.getBodyBlock());
+
+  auto &op = alwaysBlocks[{builder.getBlock(), clockEdge, clock, resetStyle,
+                           resetEdge, reset}];
+  auto &alwaysOp = op.first;
+  auto &insideIfOp = op.second;
+
+  if (!alwaysOp) {
+    if (reset) {
+      assert(resetStyle != ::ResetType::NoReset);
+      // Here, we want to create the folloing structure with sv.always and
+      // sv.if. If `reset` is async, we need to add `reset` to a sensitivity
+      // list.
+      //
+      // sv.always @(clockEdge or reset) {
+      //   sv.if (reset) {
+      //     resetBody
+      //   } else {
+      //     body
+      //   }
+      // }
+
+      auto createIfOp = [&]() {
+        // It is weird but intended. Here we want to create an empty sv.if
+        // with an else block.
+        insideIfOp = builder.create<sv::IfOp>(
+            reset, []() {}, []() {});
+      };
+      if (resetStyle == ::ResetType::AsyncReset) {
+        sv::EventControl events[] = {clockEdge, resetEdge};
+        Value clocks[] = {clock, reset};
+
+        alwaysOp = builder.create<sv::AlwaysOp>(events, clocks, [&]() {
+          if (resetEdge == sv::EventControl::AtNegEdge)
+            llvm_unreachable("negative edge for reset is not expected");
+          createIfOp();
+        });
+      } else {
+        alwaysOp = builder.create<sv::AlwaysOp>(clockEdge, clock, createIfOp);
+      }
+    } else {
+      assert(!resetBody);
+      alwaysOp = builder.create<sv::AlwaysOp>(clockEdge, clock);
+      insideIfOp = nullptr;
+    }
+  }
+
+  if (reset) {
+    assert(insideIfOp && "reset body must be initialized before");
+    auto resetBuilder =
+        ImplicitLocOpBuilder::atBlockEnd(loc, insideIfOp.getThenBlock());
+    resetBody(resetBuilder);
+
+    auto bodyBuilder =
+        ImplicitLocOpBuilder::atBlockEnd(loc, insideIfOp.getElseBlock());
+    body(bodyBuilder);
+  } else {
+    auto bodyBuilder =
+        ImplicitLocOpBuilder::atBlockEnd(loc, alwaysOp.getBodyBlock());
+    body(bodyBuilder);
+  }
+}
+
 void SeqToSVPass::runOnOperation() {
   ModuleOp top = getOperation();
-  MLIRContext &ctxt = getContext();
 
+  for (auto module : top.getOps<hw::HWModuleOp>())
+    FirRegLower(module).lower();
```

Comment:
Can't this be handled by the partialConversion framework?  The main issue is storing the state for the always merging outside the pattern.